### PR TITLE
Make the tests go through the no-op restore codepath

### LINKED ISSUE #10211: Add metadata to restore tests, make them go through the no-op codepath
While I investigating https://github.com/NuGet/Home/issues/10068, I hit an known problem of the no-op codepath not being exercised in a good amount of unit test. 

The reason is that the packagespec is not created with the same amount of details prod package spec would be.
In prod scenarios, all restores go through the restorerunner which calls https://github.com/NuGet/NuGet.Client/blob/c3219e49852add8371db25680a300cce5551ef00/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/SpecValidationUtility.cs#L28 to validate that things like the `RestoreMedata` are not null or empty.

Now this is not a major problem for no-op, I have not discovered any regressions because of this, but it *does* not reflect what the actual restore code run is.

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/10211
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 

While I investigating https://github.com/NuGet/Home/issues/10068, I hit an known problem of the no-op codepath not being exercised in a good amount of unit test.

The reason is that the packagespec is not created with the same amount of details prod package spec would be.
In prod scenarios, all restores go through the restorerunner which calls https://github.com/NuGet/NuGet.Client/blob/c3219e49852add8371db25680a300cce5551ef00/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/SpecValidationUtility.cs#L28 to validate that things like the RestoreMedata are not null or empty.

Now this is not a major problem for no-op, I have not discovered any regressions because of this, but it does not reflect what the actual restore code run is.

Now the TestRestoreRequest is written in a way that'd prevent restore going through unexpected codepaths. 

Note that I am aware that these tests are running project.json. 
Changing them to PR is not super straightforward and can be done as a future step. https://github.com/NuGet/Home/issues/10212

## Testing/Validation

Tests Added: No  
Reason for not adding tests: I am changing all the tests :) 
Validation:

### CODE REVIEW

**Comment by nkolev92** [comment]
@NuGet/nuget-client 🔔 
Review please. It's been ~10 days.

File: src/NuGet.Core/NuGet.LibraryModel/LibraryDependencyTypeFlag.cs [hunk 1]
```diff
@@ -40,7 +41,9 @@ public override bool Equals(object obj)
 
         public override int GetHashCode()
         {
-            return StringComparer.Ordinal.GetHashCode(_value);
+            var combiner = new HashCodeCombiner();
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.LibraryModel/LibraryDependencyTypeFlag.cs:44]: What is the benefit of changing for HashCodeCombiner in this context?
Isn't just `StringComparer.Ordinal.GetHashCode` enough for simple string?

**Review by zivkan — APPROVED** [review]
GitHub has all the files marked as viewed, meaning I already reviewed this, who knows how long ago, but forgot to submit a review 😞

File: src/NuGet.Core/NuGet.LibraryModel/LibraryDependencyTypeFlag.cs (hunk 1 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.LibraryModel/LibraryDependencyTypeFlag.cs:44]: Have a look at the remarks for the docs on `string.GetHashCode()`: https://docs.microsoft.com/en-us/dotnet/api/system.string.gethashcode?view=netcore-3.1

Particularly this bit:

> The hash code itself is not guaranteed to be stable. Hash codes for identical strings can differ across .NET implementations, across .NET versions, and across .NET platforms (such as 32-bit and 64-bit) for a single version of .NET. In some cases, they can even differ by application domain. This implies that two subsequent runs of the same program may return different hash codes.
> 
> As a result, hash codes should never be used outside of the application domain in which they were created, they should never be used as key fields in a collection, and they should never be persisted.

Classes used in no-op restore calculations have some hash persisted to disk. I'm not sure if it's the `GetHashCode` that is persisted, or some other hash, but we do persist a hash to disk to compare against to determine no-op restore or if a "full" restore is required. Hence it's critical to have a stable hash.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.LibraryModel/LibraryDependencyTypeFlag.cs:44]: It's not used for persisting yet, but the stable implementation allows us to consider it in a future work item, like say https://github.com/NuGet/Home/issues/10068

File: test/TestUtilities/Test.Utility/Commands/ProjectJsonTestHelpers.cs [hunk 2]
```diff
@@ -147,5 +161,30 @@ public static PackageSpec WithTestRestoreMetadata(this PackageSpec spec)
             }
             return updated;
         }
+
+        private static PackageSpec WithProjectJsonTestRestoreMetadata(this PackageSpec spec)
+        {
+            var updated = spec.Clone();
```
Comment (kartheekp-ms) [inline test/TestUtilities/Test.Utility/Commands/ProjectJsonTestHelpers.cs:167]: Can we skip cloning here? If am trying to understand why we need to clone here.

**Pull request merged**